querqy-for-lucene/querqy-solr/src/main/java/querqy/solr/explain/SnapshotRewriter.java

```diff
@@ -0,0 +1,237 @@
+        return query;
+    }
+
+    @Override
+    public ExpandedQuery rewrite(final ExpandedQuery query) {
+        throw new UnsupportedOperationException("This rewriter needs a query context");
+    }
+
+    public Map<String, Object> getSnapshot() {
+        return snapshot == null ? Collections.emptyMap() : snapshot;
+    }
+
+
+    public static class QuerySnapshotVisitor extends AbstractNodeVisitor<Void> {
+
+        Map<String, Object> snapshot = new HashMap<>();
+        List<Map<String,?>> clauses = new LinkedList<>();
+
+        public Map<String, Object> takeSnapshot(final QuerqyQuery<?> querqyQuery) {
+            if (querqyQuery instanceof Query) {
+                reset();
+                visit((Query) querqyQuery);
+            } else if (querqyQuery instanceof BooleanQuery) {
+                visit((BooleanQuery) querqyQuery);
+            } else if (querqyQuery instanceof MatchAllQuery) {
+                visit((MatchAllQuery) querqyQuery);
+            } else if (querqyQuery instanceof RawQuery) {
+                visit((RawQuery) querqyQuery);
+            } else {
+                throw new IllegalArgumentException("Cannot handle query type " + querqyQuery.getClass());
+            }
+
+            return snapshot;
+
+        }
+
+        @Override
+        public Void visit(final Query query) {
+            snapshot.put(TYPE_QUERY, clauses);
+            super.visit(query);
+            return null;
+        }
+
+        @Override
+        public Void visit(final BooleanQuery booleanQuery) {
+
+            final Map<String, Object> bq = new HashMap<>();
+            clauses.add(bq);
+            final Map<String, Object> data = new LinkedHashMap<>(); // preserve order
+            bq.put(TYPE_BOOLEAN_QUERY, data);
+            data.put(PROP_OCCUR, booleanQuery.occur.name());
+
+            // save reference
+            final List<Map<String,?>> origClauses = clauses;
+            clauses = new LinkedList<>();
+            data.put(PROP_CLAUSES, clauses);
+
+            super.visit(booleanQuery);
+
+            snapshot.put(TYPE_BOOLEAN_QUERY, clauses);
+            // restore
+            clauses = origClauses;
+
+            return null;
+        }
+
+        @Override
+        public Void visit(final DisjunctionMaxQuery disjunctionMaxQuery) {
+            final Map<String, Object> dmq = new HashMap<>();
+            clauses.add(dmq);
+            final Map<String, Object> data = new LinkedHashMap<>(); // preserve order
+            dmq.put(TYPE_DISMAX, data);
+            data.put(PROP_OCCUR, disjunctionMaxQuery.occur.name());
+
+            // save reference
+            final List<Map<String,?>> origClauses = clauses;
+            clauses = new LinkedList<>();
+            data.put(PROP_CLAUSES, clauses);
+
+            super.visit(disjunctionMaxQuery);
+            clauses = origClauses;
+
+            return null;
+
+        }
+
+        @Override
+        public Void visit(final Term term) {
```

Comment:
`term` could also be of type `BoostedTerm` in that case we should add the boost to the snapshot